dhis2/dhis2-core
# feat: Cascading sharing for dashboards to items

The PR introduces new api endpoint  `api/dashboards/cascadeSharing/{dashboardUID}` with POST request
- The function will take `UserAccesses` and `UserGroupAccesses` from given `Dashboard`'s sharing and add them to all `DashboardItem` objects. 
- The added `UserAccesses` and `UserGroupAccesses` will only have `METADATA_READ` permission.
- A new object `AccessObject` which is extended by `UserAccess` and `UserGroupAccess` , this help reduces the codes to handle those objects.
- If the target objects already have `publicAccess` enabled, then it will be ignored, no access objects will be added.
- Currently there are four `DashboardItemType` supported : `Map`, `Visualization`, `EventChart`, `EventReport`.
- The api has a parameter `dryRun=true/false` ( default is false ), if its set to `true` then no object will updated. 
- The api has a parameter `atomic=true/false` ( default is false ), if its set to `true` no objects will be saved if there is an error, otherwise try with best effort.
- The api will return a report as below
```
{
  "errorReports": [
    {
      "message": "No matching object for reference. Identifier was s46m5MS0hxu, and object was DataElement.",
      "mainKlass": "org.hisp.dhis.dataelement.DataElement",
      "errorCode": "E5001",
      "errorProperties": [
        "s46m5MS0hxu",
        "DataElement"
      ]
    }
  ],
  "countUpdatedDashBoardItems": 1,
  "updateObjects": {
    "dataElements": [
      {
        "id": "YtbsuPPo010",
        "name": "Measles doses given"
      },
      {
        "id": "l6byfWFUGaP",
        "name": "Yellow Fever doses given"
      }
    ]
  }
}
```

### Test
- Unit tests are added for common cases.

- Manual test case with Sierra Leone DB:
1) Create a new userA with authority view Dashboard.
2) Create a new dashboardA with no `publicAccess`. Share it with userA with METADATA_READ access
3) In dashboardA, add a `Visualization` `Immunization: BCG, measles, yellow fever last 12 months (stacked)`
4) Go to Visualization app, disabled `publicAccess` for added visualization, only share with admin user.
5) In Maintenance app, disabled `publicAccess` and only share to admin user for  those `DataElement`:     
"YtbsuPPo010",  "s46m5MS0hxu", "l6byfWFUGaP"
6) Login with userA and check dashboardA, **no data should be displayed.**
7) Using admin user, send POST request to `api/dashboards/cascadeSharing/{dashboardUID}`
8) Login with userA, check dashboardA, **data should be displayed now.**
